pkg/actuators/machine/instances.go

```diff
@@ -148,9 +148,49 @@ func getSubnetIDs(machine runtimeclient.ObjectKey, subnet machinev1beta1.AWSReso
 	if len(subnetIDs) == 0 {
 		return nil, fmt.Errorf("no subnet IDs were found")
 	}
+
+	subnetIDFromAvalabilityZone, err := getSubnetIDFromAvalabilityZone(availabilityZone, client)
+	if err != nil {
+		klog.Errorf("could not check if the subnet id and availability zone fields are mismatched")
+		return subnetIDs, nil
+	}
+
+	if *subnetIDs[0] != subnetIDFromAvalabilityZone {
+		klog.Errorf("mismatched subnet id and availibity zone")
+	}
+
 	return subnetIDs, nil
 }
 
+// getSubnetIDFromAvalabilityZone gets a subnet id from specified availability zone
+func getSubnetIDFromAvalabilityZone(availabilityZone string, client awsclient.Client) (string, error) {
+	result, err := client.DescribeSubnets(&ec2.DescribeSubnetsInput{
+		DryRun: aws.Bool(true),
+		Filters: []*ec2.Filter{
+			{
+				Name: aws.String("availability-zone"),
+				Values: []*string{
+					aws.String(availabilityZone),
+				},
+			},
+		},
+	})
+	if err != nil {
+		return "", fmt.Errorf("could not describe a subnet")
+	}
+
+	if result == nil {
+		return "", fmt.Errorf("resulting subnet is not expected to be nil")
+	}
+
+	if len(result.Subnets) > 0 {
+		subnetID := *result.Subnets[0].SubnetId
+		return subnetID, nil
+	}
```

Comment:
This is likely to cause issues, I think there's guaranteed to be more than one subnet in each AZ and there's no guarantee that it will be the one requested for this cluster/machine.

Can we swap the logic around, look up the subnet based on the ID we have, and then check what AZ it is in? That will be safer since the mapping of subnet->AZ is 1:1 but AZ-> is 1:many

Reply:
Addressed in the latest commit. 
Checking the mismatch between `subnet.ID` and `availabilityZone` iff the subnet has ID specified. If it's nil, then we know that we are using a filter and this get the machine Failed anyway, if the values mismatch as it is mentioned in the bug card.